# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/test/test_mp3_broadcast.html
# Commit: f33e1ce6dcf1
# Full Hash: f33e1ce6dcf19e47952f3963c8e970c7c380acd2
# Author: Paul Adenot <paul@paul.cx>
# Date: 2023-07-20 21:19:23
# Regressor Bug: 1843003
# File Overlap Count: 2
# Description:
#   Bug 1843003 - Test that typical mp3 broadcast work well. r=alwu
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D183737
# ==============================================================================

diff -r 763b34de3c14 -r f33e1ce6dcf1 dom/media/test/test_mp3_broadcast.html
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/media/test/test_mp3_broadcast.html	Thu Jul 20 13:17:59 2023 +0000
@@ -0,0 +1,52 @@
+<!DOCTYPE HTML>
+<html>
+
+<head>
+    <title>Test playback of broadcast-like streams</title>
+    <script src="/tests/SimpleTest/SimpleTest.js"></script>
+    <link rel="stylesheet" type="text/css" href="/tests/SimpleTest/test.css" />
+    <script type="text/javascript" src="manifest.js"></script>
+</head>
+
+<body>
+    <audio controls id=a style="width: 100%;"></audio>
+    <script type="module">
+        SimpleTest.waitForExplicitFinish();
+        const streams = [
+            // An mp3 bytestream consisting of a complete mp3 file with a XING
+            // header, that has duration information, but served without a
+            // `Content-Length`. It is followed by a second mp3 bytestream that
+            // also has a XING header. While some software are able to play the
+            // entire file, Web browser don't.
+            { src: "two-xing-header-no-content-length.mp3", duration: 1 },
+            // An mp3 bytestream consisting of a complete mp3 file with a XING
+            // header, that has duration information, but served without a
+            // `Content-Length` header. It is followed by a second mp3 bytestream that
+            // doesn't have a XING header.
+            // This scenario is typical in radio broadcast scenario, when the
+            // live-stream has a pre-recorded prelude. The reported duration,
+            // after "ended" has been received, is the duration of playback.
+            { src: "single-xing-header-no-content-length.mp3", duration: 11.030997 },
+        ];
+        var audio = window.a;
+        // Prevent ESLint error about top-level await
+        (async function () {
+            for (let i of streams) {
+                audio.src = i.src;
+                audio.load();
+                audio.play();
+                audio.onerror = (e) => {
+                    ok(false, `${i}: error: ${e.message}}`);
+                };
+                await once(audio, "ended");
+                ok(true, `${i}: playback through the end`);
+                is(audio.duration, i.duration, "Duration at end is correct");
+                is(audio.currentTime, i.duration, "Current time at end is correct");
+            }
+            SimpleTest.finish();
+        })()
+    </script>
+</body>
+
+</html>
+